File: app/src/main/java/com/maksk993/cryptoportfolio/data/repository/CryptoRepositoryImpl.java
package com.maksk993.cryptoportfolio.data.repository;

import android.util.Log;

import com.maksk993.cryptoportfolio.data.models.CoinMarketCapAPI;
import com.maksk993.cryptoportfolio.data.models.CryptoCurrencies;
import com.maksk993.cryptoportfolio.domain.repository.DataReceivedCallBack;
import com.maksk993.cryptoportfolio.domain.repository.CryptoRepository;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;

import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class CryptoRepositoryImpl implements CryptoRepository {
    Retrofit retrofit = new Retrofit.Builder()
            .baseUrl("https://pro-api.coinmarketcap.com/")
            .addConverterFactory(GsonConverterFactory.create())
            .build();
    private static final String API_KEY = "API_KEY"; // Insert your key

    @Override
    public void getData(DataReceivedCallBack callBack) {
        CoinMarketCapAPI coinMarketCapApi = retrofit.create(CoinMarketCapAPI.class);

        for (CryptoCurrencies currency : CryptoCurrencies.values()) {
            int id = currency.getCoinMarketCapId();
            Call<ResponseBody> call = coinMarketCapApi.getCurrencyInfo(API_KEY, id);
            call.enqueue(new Callback<ResponseBody>() {
                @Override
                public void onResponse(Call<ResponseBody> call, Response<ResponseBody> response) {
                    if (response.isSuccessful()) {
                        try {
                            JSONObject jsonObject = new JSONObject(response.body().string());
                            JSONObject currencyData = jsonObject.getJSONObject("data").getJSONObject(Integer.toString(id));
                            float price = (float) currencyData.getJSONObject("quote").getJSONObject("USD").getDouble("price");
                            String symbol = currencyData.getString("symbol");
                            callBack.dataReceived(symbol, price);
                            // Log.d("PRICES", currencyData.toString());
                        } catch (JSONException | IOException e) {
                            Log.d("PRICES", e.toString());
                        }
                    } else {
                        Log.d("PRICES", "response is not successful.");
                    }
                }

                @Override
                public void onFailure(Call<ResponseBody> call, Throwable t) {
                    Log.d("PRICES", "response is not successful.");
                }
            });
        }
    }
}

File: app/src/main/java/com/maksk993/cryptoportfolio/data/repository/CryptoRepositoryImpl.kt
package com.maksk993.cryptoportfolio.data.repository

import android.util.Log
import com.maksk993.cryptoportfolio.BuildConfig
import com.maksk993.cryptoportfolio.data.models.retrofit.CryptoServiceAPI
import com.maksk993.cryptoportfolio.domain.models.Asset
import com.maksk993.cryptoportfolio.domain.repository.CryptoRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory


class CryptoRepositoryImpl : CryptoRepository {
    private val retrofit: Retrofit = Retrofit.Builder()
        .baseUrl("https://pro-api.coinmarketcap.com/")
        .addConverterFactory(GsonConverterFactory.create())
        .build()
    private val apiKey: String = BuildConfig.API_KEY
    private val service: CryptoServiceAPI = retrofit.create(CryptoServiceAPI::class.java)

    companion object {
        val actualPrices: MutableMap<String, Float> = HashMap()
    }

    override suspend fun getPrices(): Flow<Asset> {
        return flow {
            service.getPrices(apiKey).data.forEach { assetJson ->
                val asset = Asset(
                    assetJson.symbol,
                    assetJson.quote.usd.price
                )
                emit(asset)
                actualPrices[asset.symbol] = asset.price
            }
        }
            .flowOn(Dispatchers.IO)
            .catch { Log.d("PRICES", "FAIL") }
    }
}